File: studio-web/src/modelRegistry/contracts.ts
/** Production Dock — frontend types mirroring shared contracts (frozen M42). */

export type PreferenceScope = "system" | "user" | "project";

export type ProductionRuntimeSource = "local" | "api" | "hybrid";

export type CpuFallbackPolicy = "disabled" | "ask" | "lightweight_only";

export type CapabilityLabel =
  | "Certified"
  | "Testing"
  | "Available"
  | "Unavailable"
  | "Unsupported"
  | "Requires Setup"
  | "Loading"
  | "Error";

export type LocalLifecycle =
  | "Installed"
  | "Loading"
  | "Loaded"
  | "Unloading"
  | "VRAM insufficient"
  | "Worker offline"
  | "Dependency error"
  | "CPU-only framework detected";

export type Modality = "llm" | "video" | "image" | "audio";

export type ModelRoutingPreferenceKind =
  | "local_preferred"
  | "hosted_preferred"
  | "ask_before_switching"
  | "manual_only";

export type ThemePreference = "aurora-night" | "aurora-day" | "system";

export type HostedProviderId = "kie" | "wavespeed" | "fal" | "automatic";

export type ExecutionClass = "native_local" | "docker_local" | "hosted_api";

export interface ModelDescriptor {
  id: string;
  modality: Modality;
  label: string;
  locality: "local" | "hosted";
  /** W47: Native Local / Docker Local / Hosted API honesty */
  executionClass?: ExecutionClass;
  runtimeId?: string;
  providerId?: string;
  capabilityLabel: CapabilityLabel;
  lifecycle?: LocalLifecycle;
  supports: string[];
  doesNotSupport: string[];
  estimatedVramGb?: number;
  gpuCompatible: boolean;
  executable: boolean;
  selectable?: boolean;
  readiness?: string;
  providerModelId?: string;
}

export interface DiscoveredApiModel extends ModelDescriptor {
  readiness?: string;
  selectable?: boolean;
  accountAccessible?: boolean;
  adapterAvailable?: boolean;
}

export interface ApiModelsSectionMeta {
  activeProviderId?: string | null;
  emptyReason?: string | null;
  emptyMessage?: string | null;
  summary?: Record<string, unknown>;
  updatedAt?: string | null;
}

export interface ModalityModelSections {
  local: ModelDescriptor[];
  api: DiscoveredApiModel[];
  apiMeta?: ApiModelsSectionMeta;
}

export interface ModelRoutingPreference {
  modality: Modality;
  preference: ModelRoutingPreferenceKind;
  availableModelIds: string[];
  activeModelId?: string | null;
  fallbackModelId?: string | null;
  allowFallback: boolean;
}

export interface PreferenceProvenance {
  activeModelId?: string | null;
  activeLabel: string;
  source: PreferenceScope;
  fallbackPolicy: string;
  gpu: "Ready" | "Unavailable" | "Unknown";
  executable: boolean;
  blockedReason?: string | null;
  runtime?: string | null;
  providerId?: string | null;
  cpuFallbackPolicy: CpuFallbackPolicy;
}

export interface ResolvedSelection {
  modality: Modality;
  activeModelId?: string | null;
  activeLabel: string;
  availableModelIds: string[];
  fallbackModelId?: string | null;
  allowFallback: boolean;
  provenance: PreferenceProvenance;
}

export interface ActiveModelSelection {
  modality: Modality;
  activeModelId?: string | null;
  availableModelIds: string[];
  fallbackModelId?: string | null;
  allowFallback: boolean;
  provenance: Pick<PreferenceProvenance, "source" | "executable" | "gpu"> &
    Partial<PreferenceProvenance>;
}

export interface UserGlobalPreferences {
  theme: ThemePreference;
  dockCollapsed: boolean;
  dockAutoCollapse: boolean;
  runtimeSource: ProductionRuntimeSource;
  runtimeLocalEnabled: boolean;
  runtimeApiEnabled: boolean;
  defaultHostedProviderId: HostedProviderId;
  cpuFallbackPolicy: CpuFallbackPolicy;
  llmRouting?: ModelRoutingPreference;
  videoRouting?: ModelRoutingPreference;
  imageRouting?: ModelRoutingPreference;
  audioRouting?: ModelRoutingPreference;
}

export interface ProjectPreferences {
  projectId: string;
  activeVideoModelId?: string | null;
  activeImageModelId?: string | null;
  activeAudioModelId?: string | null;
  activeLlmModelId?: string | null;
  resolution?: string | null;
  quality?: string | null;
  generatorLocks?: Record<string, unknown>;
}

export interface ProductionControlStatus {
  ok: boolean;
  runtimeSource: ProductionRuntimeSource;
  localAvailable: boolean;
  apiAvailable: boolean;
  hostedProviderConnected: boolean;
  activeHostedProviderId?: string | null;
  modalities: Partial<Record<Modality, ResolvedSelection>>;
  queueSummary?: {
    gpuQueued: number;
    hostedQueued: number;
    active: number;
  };
  healthRows?: ProductionHealthRow[];
}

export interface ProductionHealthRow {
  id: string;
  label: string;
  status: CapabilityLabel | string;
  detail?: string | null;
}

export interface ProductionControlGate {
  productionDockGo: boolean;
  requiredFlags: Record<string, boolean>;
  message?: string | null;
}

export interface ProductionQueueJob {
  id: string;
  label: string;
  status: string;
  kind?: string | null;
  projectId?: string | null;
  modality?: Modality | string | null;
  createdAt?: string | null;
  retryable?: boolean;
  cancellable?: boolean;
  workspaceRoute?: string | null;
}

export interface ProductionQueueSnapshot {
  jobs: ProductionQueueJob[];
  gpuQueued: number;
  hostedQueued: number;
  active: number;
}

export interface ProviderSwitchPreview {
  fromProviderId?: string | null;
  toProviderId: string;
  impactSummary: string;
  affectedModalities: Modality[];
  requiresConfirmation: boolean;
}

export interface ProviderSwitchConfirmBody {
  token: string;
  confirmed: boolean;
}

export type ProductionDockMenu =
  | "llm"
  | "video"
  | "image"
  | "audio"
  | "models"
  | "settings"
  | "diagnostics"
  | "queue"
  | null;

File: studio-api/app/production_control/contracts.py
"""Frozen M42 Production Dock shared contracts — see docs/release-gate/m42/PRODUCTION_DOCK_SHARED_CONTRACTS.md."""

from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field

PreferenceScope = Literal["system", "user", "project"]
ProductionRuntimeSource = Literal["local", "api", "hybrid"]
CpuFallbackPolicy = Literal["disabled", "ask", "lightweight_only"]
CapabilityLabel = Literal[
    "Certified",
    "Testing",
    "Available",
    "Unavailable",
    "Unsupported",
    "Requires Setup",
    "Loading",
    "Error",
]
LocalLifecycle = Literal[
    "Installed",
    "Loading",
    "Loaded",
    "Unloading",
    "VRAM insufficient",
    "Worker offline",
    "Dependency error",
    "CPU-only framework detected",
]
Modality = Literal["llm", "video", "image", "audio"]
RoutingPreference = Literal[
    "local_preferred",
    "hosted_preferred",
    "ask_before_switching",
    "manual_only",
]
ThemePreference = Literal["aurora-night", "aurora-day", "system"]
GpuStatus = Literal["Ready", "Unavailable", "Unknown"]


ExecutionClass = Literal["native_local", "docker_local", "hosted_api"]


class ModelDescriptor(BaseModel):
    id: str
    modality: Modality
    label: str
    locality: Literal["local", "hosted"]
    """W47 additive: Native Local / Docker Local / Hosted API honesty."""
    executionClass: Optional[ExecutionClass] = None
    runtimeId: Optional[str] = None
    providerId: Optional[str] = None
    capabilityLabel: CapabilityLabel
    lifecycle: Optional[LocalLifecycle] = None
    supports: list[str] = Field(default_factory=list)
    doesNotSupport: list[str] = Field(default_factory=list)
    estimatedVramGb: Optional[float] = None
    gpuCompatible: bool = False
    executable: bool = False


class ModelRoutingPreference(BaseModel):
    modality: Modality
    preference: RoutingPreference = "local_preferred"
    availableModelIds: list[str] = Field(default_factory=list)
    activeModelId: Optional[str] = None
    fallbackModelId: Optional[str] = None
    allowFallback: bool = False


class PreferenceProvenance(BaseModel):
    activeModelId: Optional[str] = None
    activeLabel: str = ""
    source: PreferenceScope = "system"
    fallbackPolicy: str = "disabled"
    gpu: GpuStatus = "Unknown"
    executable: bool = False
    blockedReason: Optional[str] = None
    runtime: Optional[str] = None
    providerId: Optional[str] = None
    cpuFallbackPolicy: CpuFallbackPolicy = "disabled"


class ResolvedSelection(BaseModel):
    modality: Modality
    activeModelId: Optional[str] = None
    activeLabel: str = ""
    source: PreferenceScope = "system"
    fallbackPolicy: str = "disabled"
    gpu: GpuStatus = "Unknown"
    executable: bool = False
    blockedReason: Optional[str] = None
    runtime: Optional[str] = None
    providerId: Optional[str] = None
    cpuFallbackPolicy: CpuFallbackPolicy = "disabled"
    availableModelIds: list[str] = Field(default_factory=list)
    fallbackModelId: Optional[str] = None
    allowFallback: bool = False
    provenance: PreferenceProvenance = Field(default_factory=PreferenceProvenance)


class UserGlobalPreferences(BaseModel):
    theme: ThemePreference = "aurora-night"
    dockCollapsed: bool = False
    dockAutoCollapse: bool = True
    runtimeSource: ProductionRuntimeSource = "hybrid"
    defaultHostedProviderId: Optional[str] = None
    cpuFallbackPolicy: CpuFallbackPolicy = "disabled"
    llm: ModelRoutingPreference = Field(
        default_factory=lambda: ModelRoutingPreference(modality="llm")
    )
    video: ModelRoutingPreference = Field(
        default_factory=lambda: ModelRoutingPreference(modality="video")
    )
    image: ModelRoutingPreference = Field(
        default_factory=lambda: ModelRoutingPreference(modality="image")
    )
    audio: ModelRoutingPreference = Field(
        default_factory=lambda: ModelRoutingPreference(modality="audio")
    )
    defaultVideoQuality: str = "balanced"
    defaultImageQuality: str = "balanced"
    defaultAudioQuality: str = "balanced"


class ProjectPreferences(BaseModel):
    projectId: str
    activeVideoModelId: Optional[str] = None
    activeImageModelId: Optional[str] = None
    activeAudioModelId: Optional[str] = None
    codirectorModelId: Optional[str] = None
    resolution: Optional[str] = None
    videoQuality: Optional[str] = None
    imageQuality: Optional[str] = None
    audioQuality: Optional[str] = None
    generatorLocks: dict[str, bool] = Field(default_factory=dict)


class ActiveModelSelection(BaseModel):
    modality: Modality
    activeModelId: Optional[str] = None
    availableModelIds: list[str] = Field(default_factory=list)
    fallbackModelId: Optional[str] = None
    allowFallback: bool = False
    provenance: PreferenceProvenance = Field(default_factory=PreferenceProvenance)
